source/blog/2021-03-29-dbal-2.13.md

```diff
@@ -0,0 +1,142 @@
+deprecation going forward and settled on introducing a tiny abstraction on
+top of `@trigger_error()`.
+
+The reason for this abstraction is the potential for side effects caused by an
+error handler and the potential overhead. We expect our deprecations to be
+triggered a few hundred times in some requests as such the production overhead
+must be minimal. 
+
+This means deprecation logging is disabled by default and you must enable
+it to either use `@trigger_error` or a PSR-3 compatible logger.
+
+See the [deprecation library
+README.md](https://github.com/doctrine/deprecations/) for details on how to
+configure and use it.
+
+### PHP 7.1 and 7.2 Support
+
+A few large Doctrine DBAL deployments still support older versions of PHP that
+are not officially supported anymore, but are covered by support of a few Linux
+distributions.
+
+To provide the largest possible flexibility to the ecosystem to run code with
+both DBAL 2 or 3 this version of Doctrine DBAL will work again with PHP 7.1 and
+7.2
+
+### Migrate to DBAL 3
+
+We recommend a three step strategy to move your code-base and that of your
+dependents to DBAL 3. It depends if you are working on a library or
+platform that is dependent upon, or if you are working on a standalone application.
+
+For a standalone application:
+
+1. Upgrade to DBAL 2.13 and enforce "^2.13" as a version constraint in
+   composer.json
+2. Enable deprecation tracking and eliminate all deprecations triggered in your
+   codebase.
+3. After fixing all deprecations, update composer constraint to "^3.0".
+   Doctrine will upgrade to version 3 if all other dependencies you are using
+   are ready as well to upgrade to version 3. If it fails, you need to identify
+   and update the dependencies as well.
+
+For a library, framework or platform:
+
+1. Upgrade to DBAL 2.13 and enforce "^2.13" as a version constraint in
+   composer.json
+2. Enable deprecation tracking and eliminate all deprecations triggered in your
+   codebase. Release a version so that all plugins and downstream users
+   can be notified of using deprecated Doctrine DBAL directly themselves.
+3. After fixing all deprecations, update composer constraint to "^2.13 | ^3.0".
+   Doctrine will only upgrade to version 3 if all other dependencies you are
+   using are ready as well to upgrade to version 3.
+
+We recommend the following strategies to detect the use of deprecated code:
+
+1. If you have an extensive test-suite, register a PSR-3 logger with Doctrine
+   Deprecations to catch all deprecations while running the tests.
+2. Otherwise register a PSR-3 logger with Doctrine Deprecations in development
+   or staging only and collect and fix them as well you can. This could be
+   done for a longer amount of time of days, weeks or months.
+3. If that is not possible, register a PSR-3 logger in production. Make sure to
+   eliminate high frequency deprecations quickly or call `ignoreDeprecation` to
+   snooze them to avoid overhead.
```

Comment:
It may be worth mentioning Psalm which reports the usage of deprecated symbols at level 2. Not sure if it's adopted at this level by the consumers of the DBAL but it's one of the most reliable options (doesn't depend on test coverage, doesn't require any changes in production).